onnxruntime/core/providers/webnn/builders/impl/cast_op_builder.cc

```diff
@@ -80,22 +80,29 @@ Status CastOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
 }
 
 // Operator support related.
+bool CastOpBuilder::HasSupportedInputsImpl(const Node& node, const emscripten::val& wnn_limits,
+                                           const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& op_type = node.OpType();
+  int32_t input_type;
 
-bool CastOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& /* initializers */,
-                                      const Node& node,
-                                      const WebnnDeviceType device_type,
-                                      const logging::Logger& logger) const {
-  NodeAttrHelper helper(node);
-  // Check cast output type.
-  const auto to_type = helper.Get("to", ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED);
+  if (!GetType(*input_defs[0], input_type, logger))
+    return false;
 
-  // WebNN CPU backend doesn't support casting to uint64 data type.
-  if (device_type == WebnnDeviceType::CPU && to_type == ONNX_NAMESPACE::TensorProto_DataType_UINT64) {
-    LOGS(logger, VERBOSE) << "Cast to uint64 is not supported for WebNN CPU backend.";
+  if (!IsSupportedDataType(input_type, wnn_limits["cast"]["input"]["dataTypes"])) {
+    LOGS(logger, VERBOSE) << "[" << op_type
+                          << "] Input type: [" << input_type
+                          << "] is not supported for now";
     return false;
   }
-  if (!IsSupportedDataType(to_type, webnn_supported_data_types)) {
-    LOGS(logger, VERBOSE) << "WebNN doesn't support casting to type " << to_type << ".";
+
+  NodeAttrHelper helper(node);
+  // Check cast to type.
+  const auto to_type = helper.Get("to", ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED);
+  if (!IsSupportedDataType(to_type, wnn_limits["cast"]["output"]["dataTypes"])) {
```

Comment:
Would it be useful to have an overload of `IsSupportedDataType` that did the lookup and logged the message, rather than each caller doing the lookup and LOGS'ing it?

e.g. `!IsSupportedDataType(to_type, wnn_limits, "cast", "output")`

Reply:
Good idea, I'd propose an `IsDataTypeSupportedByOp` to check if the input or output data type of ONNX node is supported by webnn op.